### PR TITLE
fix(grpo): enable runtime-probed MPS BF16

### LINKED ISSUE #567: Enable runtime-probed MPS BF16 for local Transformers GRPO/RLVR
## Hardware reproduction

On Apple Silicon MPS with the current supported stack (PyTorch 2.12.1, Transformers 5.12.1, TRL 0.29.1), a local Transformers GRPO run with `HuggingFaceTB/SmolLM2-135M-Instruct`, LoRA, `num_generations: 2`, and deterministic math RLVR completes successfully after the reward-data fix in #566.

However, `GRPOTrainerWrapper._build_precision_kwargs()` still hard-codes both precision flags to false for every non-CUDA device. The measured run therefore reports:

```text
device=mps
bf16=false
fp16=false
frozen base parameters=torch.bfloat16
trainable LoRA parameters=torch.float32
```

This differs from the MPS runtime-probed BF16 policy added for the validated SFT, DPO, reward-model, and PRM trainers in #564. GRPO's local Transformers path uses the same causal-LM/LoRA loading primitives and does not require vLLM.

## Proposed fix

- opt the local Transformers GRPO trainer into `bf16_fp16_flags(..., allow_mps_bf16=True)`
- retain FP32 fallback when the live MPS runtime rejects a BF16 allocation
- keep CPU, CUDA, vLLM, quantization, and MLX behavior unchanged
- add unit coverage for both MPS capability outcomes
- run a hardware smoke with deterministic RLVR, verify finite training state, save the adapter, and reload it

The implementation will be based on #564 and #566, so the resulting PR will be explicitly marked as dependent on both until they merge.

## Ownership

I am taking this issue and will prepare the focused fix and Apple Silicon evidence.

### PR DESCRIPTION
**Pull request by Amix29**
## Dependencies

Draft until both prerequisite PRs merge:

- #564 provides the shared runtime-probed MPS BF16 policy and validates it in the resident text trainers.
- #566 fixes the GRPO data/reward contract needed for the documented local Alpaca/RLVR path to reach training.

The focused change for this PR is commit `a095405`. I will rebase it onto `main` after both dependencies merge.

## Summary

- opt local Transformers GRPO/RLVR into BF16 only when the live MPS runtime accepts BF16
- retain the existing FP32 fallback for unsupported MPS runtimes and leave CPU/CUDA behavior unchanged
- keep MPS on local Transformers generation (`use_vllm: false`)
- document the newly hardware-validated surface and cover capability success, capability failure, `mps:0`, CPU, and the CUDA-only `grpo_fp16` override

## Apple Silicon hardware evidence

Tested offline with PyTorch 2.12.1, Transformers 5.12.1, TRL 0.29.1, and `HuggingFaceTB/SmolLM2-135M-Instruct`:

- local Transformers GRPO, LoRA rank 4, two generations, deterministic math RLVR
- process started with `PYTORCH_ENABLE_MPS_FALLBACK=0` before Torch import; Accelerate later rewrites the environment variable to `1`, but the MPS runtime was initialized with fallback disabled
- `bf16=true`, `fp16=false`, `use_vllm=false`
- 134,515,008 frozen base parameters in BF16; 230,400 trainable LoRA parameters in FP32
- one training step completed in 1.445 seconds (`train_runtime=1.573s`)
- finite loss `0.0001353885`, gradient norm `0.00201798`, KL `0.00135388`, and entropy `1.75193`
- MPS current allocation: 270,973,696 -> 296,548,864 bytes; driver allocation: 1,166,508,032 -> 1,222,524,928 bytes
- adapter save completed; reload found 120 finite LoRA tensors / 230,400 parameters
- a reloaded BF16 forward produced finite BF16 logits on `mps:0` with shape `[1, 7, 49152]`

This is a compatibility smoke, not a quality or throughput benchmark; the two generated samples scored zero reward, which is expected to be noisy for an untrained 135M model.

## Validation

- `ruff check src/soup_cli/ tests/`
- `18,957 passed, 82 skipped, 5 deselected` in the full non-smoke test suite
- 141 focused MPS precision, GRPO, RLVR, reward-metadata, and existing GRPO compatibility tests
- hardware train/save/reload/forward smoke described above

Fixes #567

### CODE REVIEW

**Comment by Amix29** [comment]
Ported the #566 multi-turn GRPO correction into this dependent branch in cec496a. Earlier assistant turns now remain in the prompt and only the final assistant reference is removed.

Validation: 60 focused GRPO, reward-metadata, and MPS BF16 tests passed; Ruff is clean.

**Comment by MakazhanAlpamys** [comment]
Thanks for the draft implementation. I reviewed the linked issues first, the full diff, test deletions, docs/changelog, and the dependency on #566.

Before this is ready for merge:
- Rebase onto current main; the branch is based on 6c13c44 and conflicts with the current tree in docs/backends-and-ops.md, docs/performance-and-quantization.md, and tests/test_issue563_mps_bf16_training.py.
- Run fresh CI after the rebase. The published checks are green for the old base, but they do not validate the merged result.
- Please update the stale Transformers-floor comment in .github/workflows/ci.yml.

Verification on the reviewed head cec496a37a4d5632e1eebe129be13bcb8773842d:
- 70 focused MPS/GRPO tests passed.
- 87 compatibility tests passed with 1 skip.
- Full Ruff passed.
- The mutation forcing bf16=False failed test_grpo_enables_bf16_when_live_mps_probe_succeeds, so the main regression test is effective.

The live MPS hardware smoke was not reproducible on this host (mps_available=False); the mocked capability path is covered. The upstream Transformers int32/int64 scatter failure observed in the Qwen4 acceptance environment is separate from this Soup change, but the compatibility result should be documented or gated if the supported floor still includes that environment.

**Comment by MakazhanAlpamys** [comment]
The prerequisite PRs #564 and #566 are now merged. Please rebase this draft onto current `main`, resolve the documented conflicts, and refresh the full CI matrix. Once the rebase is clean, mark the PR ready for review; the Apple Silicon smoke evidence and the CPU capability regressions can then be reviewed against the composed tree. Issue #567 remains open for this follow-up.

**Comment by Amix29** [comment]
Rebased onto current `main` after #564 and #566 merged. Git dropped the duplicated multi-turn GRPO patch because its contents are already upstream, leaving only the two #568-specific commits. The current Transformers-floor/Qwen4 compatibility comment is inherited from `main`, so the stale comment noted in the previous review is no longer present.

Validation on the rebased head:

- `ruff check src/soup_cli/ scripts/ tests/` — passed
- 106 focused MPS BF16, GRPO metadata/history, Transformers-floor, and Qwen4 compatibility tests passed

The PR is now marked ready for review and fresh GitHub CI is running.

**Review by MakazhanAlpamys — APPROVED** [review]
Re-reviewed current head 6cf23225 against issue #567. The local Transformers GRPO path now uses the shared runtime MPS BF16 probe, keeps CPU/CUDA/vLLM behavior unchanged, and preserves the CUDA-only grpo_fp16 override. The diff retains the existing tests and adds capability-success/failure, mps:0, CPU, and override coverage; changelog and backend/performance docs are present.

Independent verification:
- pytest tests/test_issue567_grpo_mps_bf16.py tests/test_issue563_mps_bf16_training.py tests/test_v0533.py --no-cov -q (58 passed)
- Ruff on the changed trainer/tests and git diff --check (passed)
- mutation: forcing allow_mps_bf16=False made both MPS success assertions fail (2 failed), so the tests catch removal of the main fix.

All 13 checks reported by `gh pr checks 568` are green on 6cf23225. Approving for squash merge.

**Comment by MakazhanAlpamys** [comment]
Thank you, @Amix29, for carrying the MPS BF16 policy into local Transformers GRPO with real Apple Silicon evidence and a conservative fallback. I independently ran 58 focused tests, Ruff, and `git diff --check`; forcing `allow_mps_bf16=False` made both MPS success assertions fail. All 13 current-head CI checks passed before the squash merge as `07db7a9`.
